src/common_modules/vsts-cli-package-common/vsts/cli/package/common/artifacttool_updater.py

```diff
@@ -0,0 +1,111 @@
+
+        # Call the auto-update API to find the current version of ArtifactTool
+        # If VSTS_ARTIFACTTOOL_OVERRIDE_URL is set, instead always download from the URL
+        artifacttool_override_url = os.environ.get(self.ARTIFACTTOOL_OVERRIDE_URL_ENVKEY)
+        if artifacttool_override_url is not None:
+            release_uri = artifacttool_override_url
+            release_id = "custom_%s" % str(uuid.uuid4()) # ensures that the custom URL is always downloaded fresh
+        else:
+            override_version = os.environ.get(self.ARTIFACTTOOL_OVERRIDE_VERSION_ENVKEY)
+            try:
+                release = self._get_current_release(team_instance, override_version)
+            except Exception as ex:
+                raise CLIError('Failed to find UPack tooling: %s' % ex) 
+            release_uri, release_id = release
+
+        # Determine the path for the release, and skip downloading if it already exists
+        logger.debug("Checking if we already have ArtifactTool release '%s'", release_id)
+        release_dir = self._get_release_dir(release_id)
+        if os.path.exists(release_dir):
+            logger.debug("Not updating ArtifactTool because the current release already exists at '%s'" % release_dir)
+            return release_dir
+              
+        # Doesn't already exist. Download and extract the release.
+        logger.debug("Updating to ArtifactTool release %s since it doesn't exist at %s" % (release_id, release_dir))
+        with humanfriendly.AutomaticSpinner(label="Updating UPack tooling (%s)" % release_id, show_time=False) as spinner:
+            self._download_artifacttool(release_uri, release_dir)
+        return release_dir
+
+    def _get_current_release(self, team_instance, override_version):
+        connection = get_vss_connection(team_instance)
+        client = connection.get_client('vsts.cli.package.common.client_tool.client_tool_client.ClientToolClient')
+        logger.debug("Looking up current version of ArtifactTool...")
+        release = client.get_clienttool_release("ArtifactTool", os_name=platform.system(), arch=platform.machine(), version=override_version)
+        return (release.uri, self._compute_id(release)) if release is not None else None
+
+    def _download_artifacttool(self, uri, release_dir):
+        logger.debug("Downloading ArtifactTool from %s" % uri)
+        response = requests.get(uri)
+
+        release_temp_dir = self._get_temp_release_dir()
+        logger.debug("Extracting ArtifactTool to %s" % release_temp_dir)
+        f = zipfile.ZipFile(io.BytesIO(response.content))
+        f.extractall(path=release_temp_dir)
+
+        logger.debug("Moving downloaded ArtifactTool from %s to %s" % (release_temp_dir, release_dir))
+        os.rename(release_temp_dir, release_dir)
+        logger.info("Downloaded UPack tooling successfully")
+
+    def _mkdir_if_not_exist(self, path):
+        try: 
+            os.makedirs(path)
```

Comment:
But then I'd need to essentially re-write os.makedirs, since I'd need to check if the path exists at each level. I'm happy to switch this to something better, but a quick internet search showed this as the most common approach.